src/librustc_target/abi/mod.rs

```diff
@@ -1081,4 +1119,57 @@ impl<'a, Ty> TyLayout<'a, Ty> {
             Abi::Aggregate { sized } => sized && self.size.bytes() == 0
         }
     }
+
+    /// Determines if zero-initializing this type might be okay.
+    /// This is conservative: in doubt, it will answer `true`.
+    pub fn might_permit_zero_init<C, E>(
+        &self,
+        cx: &C,
+    ) -> Result<bool, E>
+    where
+        Self: Copy,
+        Ty: TyLayoutMethods<'a, C>,
+        C: LayoutOf<Ty = Ty, TyLayout: MaybeResult<Self, Error = E>>
+    {
+        fn scalar_allows_zero(s: &Scalar) -> bool {
+            (*s.valid_range.start() <= 0) || // `&& *s.valid_range.end() >= 0` would be redundant
+            (*s.valid_range.start() > *s.valid_range.end()) // wrap-around allows 0
+        }
+
+        // Abi is the most informative here.
+        let res = match &self.abi {
+            Abi::Uninhabited => false, // definitely UB
+            Abi::Scalar(s) => scalar_allows_zero(s),
+            Abi::ScalarPair(s1, s2) =>
+                scalar_allows_zero(s1) && scalar_allows_zero(s2),
+            Abi::Vector { element: s, count } =>
+                *count == 0 || scalar_allows_zero(s),
+            Abi::Aggregate { .. } => {
+                // For aggregates, recurse.
+                let inner = match self.variants {
+                    Variants::Multiple { .. } => // FIXME: get variant with "0" discriminant.
+                        return Ok(true),
+                    Variants::Single { index } => self.for_variant(&cx, index),
+                };
+
+                match inner.fields {
+                    FieldPlacement::Union(..) => true, // An all-0 unit is fine.
+                    FieldPlacement::Array { .. } =>
+                        // FIXME: The widely use smallvec 0.6 creates uninit arrays
+                        // with any element type, so let us not (yet) complain about that.
```

Comment:
This probably warrants a tracking issue (and if `panic_if_non_zero` has a tracking issue, it could be used for all of these `FIXME`s)

Reply:
A know tracking issues for unstable features, but this is not one... so how would that look like?

Reply:
Oh right, I'm not sure. Maybe @oli-obk knows how to handle this.

Reply:
There's no predefined structure for such issues. As long as the code references the issue id so we can find the relevant code snippet even after refactorings, the issue itself needs little information.

Reply:
Okay, opened a tracking issue at https://github.com/rust-lang/rust/issues/66151. I will add a reference in the code.